File: programmers/level2/캐시.java
import java.util.*;

class Solution {
	public int solution(int cacheSize, String[] cities) {
		int answer = 0;
		for(int i = 0; i < cities.length; ++i) {
			cities[i] = cities[i].toLowerCase();
		}
		LinkedList<String> cache = new LinkedList<>();
		int idx;
		for (int i = 0; i < cities.length; ++i) {
			if ((idx = isHit(cache, cities[i])) != -1) {
				cache.remove(idx);
				answer++;
			} else {
				if (cacheSize > 0 && cache.size() == cacheSize) {
					cache.remove(cacheSize - 1);
				}
				answer += 5;
			}
			if(cacheSize > 0) {
				cache.add(0, cities[i]);
			}
		}
		return answer;
	}

	int isHit(List<String> cache, String city) {
		for (int i = 0; i < cache.size(); ++i) {
			if (cache.get(i).equals(city)) {
				return i;
			}
		}
		return -1;
	}
}

public class 캐시 {
	public static void main(String[] args) {
		Solution sol = new Solution();
		int cacheSize = 3;
		String[] cities = { "Jeju", "Pangyo", "Seoul", "NewYork", "LA", "Jeju", "Pangyo", "Seoul", "NewYork", "LA" };
		System.out.println(sol.solution(cacheSize, cities));
	}
}

File: programmers/python/level2/캐시.py
from collections import deque
def solution(cacheSize, cities):
    answer = 0
    cache = deque()
    for city in cities:
        city.lower()
        if is_hit(cache, city):
            answer += 1
        else:
            answer += 5
            if cacheSize > 0:
                if len(cache) >= cacheSize:
                    cache.popleft()
                cache.append(city)
    return answer

def is_hit(cache, city):
    if city in cache:
        cache.remove(city)
        cache.append(city)
        return True
    return False
    